repository/src/main/java/org/apache/atlas/accesscontrol/persona/AtlasPersonaService.java

```diff
@@ -0,0 +1,1423 @@
+
+                if (rangerPolicy == null) {
+                    RangerPolicy policy = fetchRangerPolicyByResources(atlasRangerService,
+                            context.isDataPolicy() ? RANGER_HEKA_SERVICE_TYPE.getString() : RANGER_ATLAS_SERVICE_TYPE.getString(),
+                            context.isDataMaskPolicy() ? RANGER_POLICY_TYPE_DATA_MASK : RANGER_POLICY_TYPE_ACCESS,
+                            provisionalPolicy);
+
+                    if (policy == null) {
+                        atlasRangerService.createRangerPolicy(provisionalPolicy);
+                    } else {
+                        rangerPolicy = policy;
+                        addNewPolicyItem = true;
+                    }
+                }
+
+                if (rangerPolicy != null) {
+                    boolean update = false;
+
+                    if (context.isDataMaskPolicy()) {
+                        List<RangerDataMaskPolicyItem> existingRangerPolicyItems = rangerPolicy.getDataMaskPolicyItems();
+                        List<RangerDataMaskPolicyItem> provisionalPolicyItems = provisionalPolicy.getDataMaskPolicyItems();
+
+                        update = updateMaskPolicyItem(context,
+                                existingRangerPolicyItems,
+                                provisionalPolicyItems,
+                                addNewPolicyItem);
+
+                    } else {
+
+                        List<RangerPolicyItem> existingRangerPolicyItems = context.isAllowPolicy() ?
+                                rangerPolicy.getPolicyItems() :
+                                rangerPolicy.getDenyPolicyItems();
+
+                        List<RangerPolicyItem> provisionalPolicyItems = context.isAllowPolicy() ?
+                                provisionalPolicy.getPolicyItems() :
+                                provisionalPolicy.getDenyPolicyItems();
+
+                        update = updatePolicyItem(context,
+                                existingRangerPolicyItems,
+                                provisionalPolicyItems,
+                                addNewPolicyItem);
+                    }
+
+                    if (update) {
+                        List<String> labels = rangerPolicy.getPolicyLabels();
+                        labels.add(getPersonaLabel(context.getPersona().getGuid()));
+                        labels.add(getPersonaPolicyLabel(context.getPersonaPolicy().getGuid()));
+                        rangerPolicy.setPolicyLabels(labels);
+
+                        atlasRangerService.updateRangerPolicy(rangerPolicy);
+                    }
+                }
+            } catch (AtlasBaseException e) {
+                LOG.error("Failed to update Ranger policies: {}", e.getMessage());
+            } finally {
+                LOG.info("End UpdateRangerPolicyWorker");
+            }
+
+            return ret;
+        }
+
+        private boolean updatePolicyItem(PersonaContext context,
```

Comment:
Mask policy and policy objects can be implemented in a factory pattern. The code would be much cleaner since it avoids such couplings (updatePolicyItem and updateMaskPolicyItem). But this of course is not a requirement for a change

Reply:
This is can't be decoupled since Ranger does not any common parent class for RangerDataMaskPolicyItem & RangerPolicyItem